src/inventory.rs

```diff
@@ -0,0 +1,197 @@
+use std::ops::Range;
+
+use crate::slot::{Slot, SlotId};
+
+pub trait Inventory {
+    fn get_slot(&self, slot_id: SlotId) -> &Slot;
+    fn set_slot(&mut self, slot_id: SlotId, slot: Slot);
+    fn capacity(&self) -> usize;
+
+    // TODO: `entry()` style api
+}
+
+pub trait CraftingInventory {
+    fn craft_result_slot() -> SlotId;
+    fn craft_table_slots() -> Range<SlotId>;
+}
+
+/// Represents a player's Inventory.
+#[derive(Debug, Clone)]
+pub struct PlayerInventory {
+    pub(crate) slots: [Slot; 46],
+}
+
+impl PlayerInventory {
+    /// General slots are the slots that can hold all items, including the
+    /// hotbar, excluding offhand. These slots are shown when the player is
+    /// looking at another inventory.
+    pub fn general_slots() -> Range<SlotId> {
+        9..45
+    }
+
+    pub fn hotbar_slots() -> Range<SlotId> {
+        36..45
+    }
+
+    pub fn hotbar_to_slot(hotbar_slot: i16) -> Option<SlotId> {
+        if !(0..=8).contains(&hotbar_slot) {
+            return None;
+        }
+
+        Some(Self::hotbar_slots().start + hotbar_slot)
+    }
+}
+
+impl Default for PlayerInventory {
+    fn default() -> Self {
+        Self {
+            // Can't do the shorthand because Slot is not Copy.
+            slots: [
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+                Slot::Empty,
+            ],
+        }
+    }
+}
+
+impl Inventory for PlayerInventory {
+    fn get_slot(&self, slot_id: SlotId) -> &Slot {
+        if slot_id < 0 {
```

Comment:
This checks the lower bound of `slot_id` but not the upper bound. The later array index will panic with a generic message.